spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceHandlerInterceptor.java

```diff
@@ -0,0 +1,172 @@
+				log.debug("Skipping creation of a span for error controller processing");
+			}
+			return true;
+		}
+		String spanName = spanName(handler);
+		Span span = getTracer().createSpan(SpanNameUtil.toLowerHyphen(spanName));
+		if (log.isDebugEnabled()) {
+			log.debug("Created new span " + span + " with name [" + spanName + "]");
+		}
+		addClassMethodTag(handler, span);
+		addClassNameTag(handler, span);
+		setSpanInAttribute(request, span);
+		return true;
+	}
+
+	private boolean isErrorControllerRelated(HttpServletRequest request) {
+		return getErrorController() != null && getErrorController().getErrorPath()
+				.equals(request.getRequestURI());
+	}
+
+	private void addClassMethodTag(Object handler, Span span) {
+		if (handler instanceof HandlerMethod) {
+			String methodName = SpanNameUtil.toLowerHyphen(
+					((HandlerMethod) handler).getMethod().getName());
+			getTracer().addTag(getTraceKeys().getHttp().getClassMethod(), methodName);
+			if (log.isDebugEnabled()) {
+				log.debug("Adding a method tag with value [" + methodName + "] to a span " + span);
+			}
+		}
+	}
+
+	private void addClassNameTag(Object handler, Span span) {
+		String className;
+		if (handler instanceof HandlerMethod) {
+			className = SpanNameUtil.toLowerHyphen(
+					((HandlerMethod) handler).getBeanType().getSimpleName());
+		} else {
+			className = SpanNameUtil.toLowerHyphen(handler.getClass().getSimpleName());
+		}
+		if (log.isDebugEnabled()) {
+			log.debug("Adding a class tag with value [" + className + "] to a span " + span);
+		}
+		getTracer().addTag(getTraceKeys().getHttp().getClassName(), className);
+	}
+
+	private String spanName(Object handler) {
+		if (handler instanceof HandlerMethod) {
+			return ((HandlerMethod) handler).getMethod().getName();
+		}
+		return handler.getClass().getSimpleName();
+	}
+
+	@Override
+	public void afterConcurrentHandlingStarted(HttpServletRequest request,
+			HttpServletResponse response, Object handler) throws Exception {
+		Span spanFromRequest = getSpanFromAttribute(request);
+		Span rootSpanFromRequest = getRootSpanFromAttribute(request);
+		if (log.isDebugEnabled()) {
+			log.debug("Closing the span " + spanFromRequest + " and detaching its parent " + rootSpanFromRequest + " since the request is asynchronous");
+		}
+		getTracer().close(spanFromRequest);
+		getTracer().detach(rootSpanFromRequest);
+	}
+
+	@Override
+	public void postHandle(HttpServletRequest request, HttpServletResponse response,
+			Object handler, ModelAndView modelAndView) throws Exception {
+		if (isErrorControllerRelated(request)) {
+			if (log.isDebugEnabled()) {
+				log.debug("Skipping closing of a span for error controller processing");
+			}
+			return;
+		}
+		getTracer().close(getSpanFromAttribute(request));
+	}
+
+	private Span getSpanFromAttribute(HttpServletRequest request) {
+		return (Span) request.getAttribute(TraceFilter.TRACE_HANDLED_REQUEST_ATTR);
+	}
+
+	private Span getRootSpanFromAttribute(HttpServletRequest request) {
```

Comment:
I'm trying no to do backwards incompatible changes in this branch, but will definitely address that in `master`